File: src/execution/execute_pipes.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   execute_pipes.c                                    :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jschwabe <jschwabe@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/03/29 18:44:46 by jschwabe          #+#    #+#             */
/*   Updated: 2024/03/29 18:44:47 by jschwabe         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "utils.h"
#include "execution.h"
#include "minishell.h"
#include "commands.h"
#include "struct.h"
#include <stdlib.h>
#include <sys/wait.h>
#include <sys/param.h>
#include <unistd.h>
#include "msh_signals.h"

static void	exec_last(t_shell *shell, int i, int *prevpipe, char **error_elem)
{
	pid_t	cpid;
	int		status;

	cpid = fork();
	status = 0;
	if (cpid == 0)
	{
		check_signals_child(&shell->p_termios);
		if (shell->token[i].has_redir)
			do_heredocs(&shell->token[i], prevpipe, shell->env);
		if (do_redirections(shell->token[i].cmd_args, error_elem) != 0)
			exit_error(shell, *error_elem);
		dup2(*prevpipe, STDIN_FILENO);
		close(*prevpipe);
		exit_free(shell, shell->token[i].cmd_func(shell, &shell->token[i]));
	}
	else
	{
		waitpid(cpid, &status, 0);
		close(*prevpipe);
		while (wait(NULL) > 0)
			;
		if (WIFEXITED(status))
			update_exit_status(shell, WEXITSTATUS(status));
	}
}

static void	exec_pipe(t_shell *shell, int i, int *prevpipe, char **error_elem)
{
	int		pipefd[2];
	pid_t	cpid;

	pipe(pipefd);
	cpid = fork();
	if (cpid == 0)
	{
		check_signals_child(&shell->p_termios);
		close(pipefd[0]);
		dup2(pipefd[1], STDOUT_FILENO);
		close(pipefd[1]);
		dup2(*prevpipe, STDIN_FILENO);
		close(*prevpipe);
		if (do_redirections(shell->token[i].cmd_args, error_elem) != 0)
			exit_error(shell, *error_elem);
		exit_free(shell, shell->token[i].cmd_func(shell, &shell->token[i]));
	}
	else
	{
		close(pipefd[1]);
		close(*prevpipe);
		*prevpipe = pipefd[0];
	}
}

void	execute_pipes(t_shell *shell, int token_count)
{
	int		i;
	int		prevpipe;
	char	*error_elem;

	i = -1;
	prevpipe = dup(STDIN_FILENO);
	while (++i < token_count - 1)
	{
		if (shell->token[i].has_redir && i != token_count - 1)
			do_heredocs(&shell->token[i], &prevpipe, shell->env);
		exec_pipe(shell, i, &prevpipe, &error_elem);
	}
	exec_last(shell, i, &prevpipe, &error_elem);
}

File: src/execution/execute_pipes.rs
#![allow(unsafe_op_in_unsafe_fn)]

use std::os::fd::{AsFd, AsRawFd, OwnedFd};

use super::{executor, heredoc::do_heredocs, redirections::do_redirections};
use crate::t_shell;
use nix::{
	errno::Errno,
	sys::wait::{WaitStatus, waitpid},
	unistd::{ForkResult, Pid},
};
use std::io::stdin;
fn exec_last(shell: &mut t_shell, i: usize, mut prevpipe: OwnedFd, pids: &mut Vec<Pid>) {
	match unsafe { nix::unistd::fork() } {
		Ok(ForkResult::Parent { child }) => {
			pids.push(child);
			drop(prevpipe);
			let mut last_status = shell.env.get_status();
			let mut remaining = pids.len();
			while remaining > 0 {
				match waitpid(None, None) {
					Ok(WaitStatus::Exited(waited_pid, exit_code)) => {
						if waited_pid == child {
							last_status = exit_code;
						}
						remaining -= 1;
					}
					Ok(WaitStatus::Signaled(waited_pid, signal, _)) => {
						if waited_pid == child {
							last_status = 128 + signal as i32;
						}
						remaining -= 1;
					}
					Ok(_) => remaining -= 1,
					Err(Errno::EINTR) => continue,
					Err(Errno::ECHILD) => break,
					Err(e) => {
						eprintln!("waitpid failed: {}", e);
						break;
					}
				}
			}
			shell.env.set_status(last_status);
		}
		Ok(ForkResult::Child) => {
			// Restore default SIGPIPE behavior so commands die silently on closed pipes
			unsafe {
				nix::sys::signal::signal(
					nix::sys::signal::Signal::SIGPIPE,
					nix::sys::signal::SigHandler::SigDfl,
				)
				.expect("pipe dfl handler");
			}
			if shell.token_vec[i].has_redir {
				do_heredocs(&shell.token_vec[i], &mut prevpipe, &shell.env);
			}
			if do_redirections(&mut shell.token_vec[i].cmd_args_vec).is_err() {
				panic!("failed to do redirections");
			}
			nix::unistd::dup2_stdin(&prevpipe).expect("dup2 stdin failed");
			drop(prevpipe);
			executor(&mut shell.token_vec[i], &mut shell.env);
			std::process::exit(shell.env.get_status());
		}
		Err(e) => eprintln!("fork failed: {}", e),
	}
}

fn exec_pipe(shell: &mut t_shell, i: usize, prevpipe: &mut OwnedFd, pids: &mut Vec<Pid>) {
	let pipefd = nix::unistd::pipe().expect("pipe fail");
	match unsafe { nix::unistd::fork() } {
		Ok(ForkResult::Parent { child }) => {
			pids.push(child);
			drop(pipefd.1);
			let old_prevpipe = std::mem::replace(prevpipe, pipefd.0);
			drop(old_prevpipe);
		}
		Ok(ForkResult::Child) => {
			// Restore default SIGPIPE behavior so commands die silently on closed pipes
			unsafe {
				nix::sys::signal::signal(
					nix::sys::signal::Signal::SIGPIPE,
					nix::sys::signal::SigHandler::SigDfl,
				)
				.expect("pipe dfl handler");
			}
			drop(pipefd.0);
			nix::unistd::dup2_stdout(&pipefd.1).expect("dup2 stdout failed");
			drop(pipefd.1);
			nix::unistd::dup2_stdin(prevpipe.as_fd()).expect("dup2 stdin failed");
			nix::unistd::close(prevpipe.as_raw_fd()).expect("close prevpipe after dup2");
			if do_redirections(&mut shell.token_vec[i].cmd_args_vec).is_err() {
				panic!("failed to do redirections");
			}
			executor(&mut shell.token_vec[i], &mut shell.env);
			std::process::exit(shell.env.get_status());
		}
		Err(e) => eprintln!("fork failed: {}", e),
	};
}
pub(super) fn execute_pipes(shell: &mut t_shell) {
	let mut prevpipe = nix::unistd::dup(stdin()).unwrap();
	let mut pids = Vec::with_capacity(shell.token_vec.len());
	for i in 0..shell.token_len.unwrap() - 1 {
		if shell.token_vec[i].has_redir && i != shell.token_len.unwrap() - 1 {
			do_heredocs(&shell.token_vec[i], &mut prevpipe, &shell.env);
		}
		exec_pipe(shell, i, &mut prevpipe, &mut pids);
	}
	exec_last(shell, shell.token_len.unwrap() - 1, prevpipe, &mut pids);
}
